src/value/native_reader.rs

```diff
@@ -47,24 +47,35 @@ impl ElementReader for NativeElementReader {
 }
 
 impl<R: RawReader> NativeElementIterator<R> {
-    /// Recursively materialize the next Ion value in the stream and return it as `Ok(Some(element))`.
-    /// If there are no more values at this level, returns `Ok(None)`.
-    /// If an error occurs while materializing the value, returns an `Err`.
+    /// Advances the reader to the next value in the stream and uses [Self::materialize_current]
+    /// to materialize it.
```

Comment:
🗺️ The original implementation of `materialize_next()` would advance the reader to the next item in the stream and then consume it recursively.

When the reader went to read struct fields, it would call `materialize_next()` and then try to read the field name. This failed for container values because `materialize_next()` would advance the cursor beyond the current field in the process of materializing it.

I broke `materialize_next()` into two methods: `materialize_current()`, which materializes the current item in the stream without first advancing the cursor, and `materialize_next()`, which is a convenience function that calls `reader.next()` and then `materialize_current()`. This allows `materialize_struct` to advance the reader to the next item in the stream once, read its field name, and then consume it via `materialize_current`.